File: rest-service/src/main/java/mb/lib/querier/BlastRequest.java
package mb.lib.querier;

import com.fasterxml.jackson.annotation.JsonGetter;
import mb.api.model.io.JsonKeys;
import mb.lib.model.HashID;
import org.veupathdb.lib.blast.BlastConfig;
import org.veupathdb.lib.blast.BlastTool;

public class BlastRequest
{
  private final HashID jobID;
  private final BlastTool tool;
  private final BlastConfig config;

  public BlastRequest(HashID jobID, BlastTool tool, BlastConfig config) {
    this.jobID  = jobID;
    this.tool   = tool;
    this.config = config;
  }

  @JsonGetter(JsonKeys.JobID)
  public HashID getJobID() {
    return jobID;
  }

  @JsonGetter(JsonKeys.Tool)
  public BlastTool getTool() {
    return tool;
  }

  @JsonGetter(JsonKeys.Config)
  public BlastConfig getConfig() {
    return config;
  }
}

File: rest-service/src/main/kotlin/mb/lib/query/model/BlastRequest.kt
package mb.lib.query.model

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.node.ObjectNode
import mb.api.model.io.JsonKeys
import org.veupathdb.lib.blast.BlastConfig
import org.veupathdb.lib.blast.BlastTool
import org.veupathdb.lib.hash_id.HashID
import org.veupathdb.lib.jackson.Json

data class BlastRequest(
  @JsonProperty(JsonKeys.JobID)
  override val jobID:  HashID,

  @JsonProperty(JsonKeys.Tool)
  val tool:   BlastTool,

  @JsonProperty(JsonKeys.Config)
  val config: BlastConfig,
) : BlastServerRequest {
  @JsonValue
  fun toJSON() = Json.newObject {
    put(JsonKeys.JobID, jobID.string)
    set<ObjectNode>(JsonKeys.Tool, tool.toJSON())
    set<ObjectNode>(JsonKeys.Config, config.toJSON().toJSON())!!
  }
}
